### DISCUSSION TITLE
Point Process model with interaction term

### DISCUSSION THREAD

**Discussion by IJCorreia** [inlabru-org/inlabru#249] [Q&A]
Dear Finn, 

This time I'm bringing a different issue but still related with the data analysis from https://github.com/inlabru-org/inlabru/discussions/240. For the same data, I would like to include an interaction term to the point process model. I am using a joint point process model with detectability, and the point (group) size as a mark. The interaction term I'm interested in is between two continuous environmental covariates. Below you can find the code for the type of model I would like to fit:

```
#Components
cmp.jointint <- ~ Intercept(1) + size.Intercept(1) + sigma_theta(1, prec.linear = 1) +
  mySPDE(main = geometry, model = matern) +
  size.SPDE(main = geometry, model = matern) +
  dist(main = f.dist(.data.), model = "linear") +
  inter(main = MatrixModels::model.Matrix(~ f.elev(.data.):f.slop(.data.), .data.), model = "fixed") +
  regi(main = border.utm$zone, model = "factor_contrast")

#Formulas
fml.jointint <- geometry + distance ~ Intercept +
  log(hn(distance, sigma.bru(sigma_theta))) +
  mySPDE + dist + inter + regi + log(2)

fml.jointsize <- size ~ size.Intercept + size.SPDE

#Likelihoods
lik.int <- inlabru::like(family = "cp",
                           formula = fml.jointint,
                           data = data.inlabrutrunc,
                           samplers = tran.inlabru,
                           domain = list(geometry = fmesher::fm_subdivide(meshUTM, n = 2),
                                         distance = fmesher::fm_mesh_1d(seq(0, max(bins), length.out = 30))))

lik.sizeTW <- inlabru::like(family = "tweedie",
                          formula = fml.jointsize,
                          data = data.inlabrutrunc)

#Model
jointfitint <- inlabru::bru(cmp.jointint, lik.int, lik.sizeTW,
                            options = list(control.inla = list(int.strategy = "eb"),
                                           #bru_max_iter = 1,
                                           control.predictor = list(compute = TRUE)))#, verbose = TRUE))

```
I am not sure about how to specify the interaction but adapted based on this (https://cran.r-project.org/web/packages/inlabru/vignettes/component.html) vignette. However, the model doesn’t fit, it issues the following error message:
![image](https://github.com/user-attachments/assets/de6af609-c3af-4c0d-bce4-6db143f6d8b4)

After `verbose = TRUE`
![image](https://github.com/user-attachments/assets/e6894155-dff4-41a8-82d8-8d272e945ba2)

Can you help me understand how to correctly specify the interaction term, or might it be related with something else?

Thank you for your attention,
Iúri J. F. Correia

**Comment by finnlindgren** [comment]
I don't think the MatrixModels code there will work as intended, as it doesn't know what to do with the function calls, and this is likely causing the error. To allow this kind of behaviour for point process models, we'd have to add some kind of pre-processing "hook"; this is vaguely planned as a future feature, but isn't yet available.

But for basic interaction terms (which doesn't necessarily make sense for continuous covariates; one should investigate whether the perceived need for a covariate product is a side-effect of non-linear effects of the individual covariates as well!) it's easy to construct the product values explicitly, and/or to use the `weights` argument:
```r
 comp <- ~ ... + elev_slop(f.elev(.data.) * f.slop(.data.), model = "const") + elev_slop_coef(1)
 formula <- ... ~ ... + elev_slop * elev_slop_coef
```
Alternatively, to keep the interaction _effect_ as a thing, do
```r
 comp <- ~ ... + elev_slop(1, weights = f.elev(.data.) * f.slop(.data.), model = "linear")
 formula <- ... ~ ... + elev_slop
```
and include the "main effects" as before, with `... + elev(f.elev(.data.)) + slop(f.slop(.data.))`.
_Maybe_ you need `rep(1, nrow(.data.))` instead of just `1` as the input, but I think the simple version with just `1` will work.

The "interaction" formula notation is only a shortcut that obscures what is _really_ used in the model; specifying it explicitly as above is safer, as it leaves no doubt what actually goes into the estimation. I would only use the shortcut notation for basic models with predefined fixed effects covariates and no point process observations.

**Reply by IJCorreia** [reply]
Hi, I apologize for taking so much to reply. Briefly, it worked, thank you for your help!! I also needed a better computer to fit the model!
